# Json Based Mini Project (World Bank Funded Projects)

#### Using data in file 'data/world_bank_projects.json' 

1) Find the 10 countries with most projects
<br>
2) Find the top 10 major project themes (using column 'mjtheme_namecode')
<br>
3) In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing    names filled in.

In [28]:
# Import all relevant packages
import pandas as pd
import json
from pandas.io.json import json_normalize
import numpy as np

### Let us load and take a look at the data 

In [2]:
# Load the data as Pandas dataframe
json_df= pd.read_json('world_bank_projects.json')

In [3]:
json_df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   
2  Active                     Y   
3  Active                     N   
4  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   
2    {'Percent': 46, 'Name': 'Regional integration'}   
3  {'Percent': 50, 'Name': 'Participation and civ...   
4  {'Percent': 30, 'Name': 'Export development an...   

                                      theme_namecode    themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]           65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...        54,24          0   
2  [{'code': '47', 'name': 'Regional integration'...  52,81,25,47    6060000   
3  [{'code': '57', 'name': 'Participation and civ...        59,57          0   
4  [{'code': '45', 'name': 'Export development an...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

In [4]:
json_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 50 columns):
_id                         500 non-null object
approvalfy                  500 non-null int64
board_approval_month        500 non-null object
boardapprovaldate           500 non-null object
borrower                    485 non-null object
closingdate                 370 non-null object
country_namecode            500 non-null object
countrycode                 500 non-null object
countryname                 500 non-null object
countryshortname            500 non-null object
docty                       446 non-null object
envassesmentcategorycode    430 non-null object
grantamt                    500 non-null int64
ibrdcommamt                 500 non-null int64
id                          500 non-null object
idacommamt                  500 non-null int64
impagency                   472 non-null object
lendinginstr                495 non-null object
lendinginstrtype            495 non

The dataframe json_df has 500 rows and 50 columns.

#### Question 1. Find the 10 countries with most projects 

Looking at the attributes, the variables 'countryname' and 'project_name' may be useful to find solution to question 1.
<br>
There are 500 countries and projects listed. Let's look how many unique combinations of countries and projects in the dataset.

In [5]:
# find unique values of countries and projects
json_df[['countryname', 'project_name']].nunique()

countryname     118
project_name    500
dtype: int64

There are 500 projects funded for total of 118 countries. Now let's look at the countries with the most projects.

In [6]:
# Find the 10 countries with most projects
json_df['countryname'].value_counts().sort_values(ascending=False).head(10)

People's Republic of China         19
Republic of Indonesia              19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Kingdom of Morocco                 12
Nepal                              12
People's Republic of Bangladesh    12
Republic of Mozambique             11
Africa                             11
Name: countryname, dtype: int64

#### Question 2.  Find the top 10 major project themes (using column 'mjtheme_namecode') 

Let's look at the column 'mjtheme_namecode'.

In [7]:
json_df[['mjtheme_namecode']].head()

mjtheme_namecode
0  [{'code': '8', 'name': 'Human development'}, {...
1  [{'code': '1', 'name': 'Economic management'},...
2  [{'code': '5', 'name': 'Trade and integration'...
3  [{'code': '7', 'name': 'Social dev/gender/incl...
4  [{'code': '5', 'name': 'Trade and integration'...

We can see that the values of column 'mjtheme_namecode' are json objects. Now we will load the file as text and flatten the nested data in the column.

In [10]:
#load the file as a text and flatten the nested data
json_text= json.load(open('world_bank_projects.json'))
mjtheme_df= json_normalize(json_text, 'mjtheme_namecode')
mjtheme_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1499 entries, 0 to 1498
Data columns (total 2 columns):
code    1499 non-null object
name    1499 non-null object
dtypes: object(2)
memory usage: 23.5+ KB


In [12]:
# find the top 10 major project themes
top_ten= mjtheme_df[mjtheme_df.name!=''].name.value_counts().sort_values(ascending=False)
top_ten.head(10)

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Economic management                              33
Name: name, dtype: int64

#### Question3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in

Let us look at the missing values in column 'mjtheme_namecode'.

In [14]:
blank_code=mjtheme_df[mjtheme_df.code==''].count()
print(blank_code)

code    0
name    0
dtype: int64


In [15]:
blank_names=mjtheme_df[mjtheme_df.name==''].count()
print(blank_names)

code    122
name    122
dtype: int64


We can see that there are 122 missing names. 
<br>
To create a dataframe with the missing names filled in, we will be doing two different approaches.
<br>
1. Find the code- name combination, create a dataframe with unique code and names combination and merge the two datframes on code.

In [16]:
# drop duplicates and find unique code, name combinations
df_drop_dup= mjtheme_df.drop_duplicates()
unique_mjtheme_df= df_drop_dup[df_drop_dup.name!='']
unique_mjtheme_df



code                                          name
0      8                             Human development
2      1                           Economic management
3      6         Social protection and risk management
4      5                         Trade and integration
5      2                      Public sector governance
6     11  Environment and natural resources management
8      7                   Social dev/gender/inclusion
11     4      Financial and private sector development
18    10                             Rural development
53     9                             Urban development
252    3                                   Rule of law

In [17]:
# reseting index
unique_mjtheme_df=unique_mjtheme_df.reset_index().drop('index', axis=1)
unique_mjtheme_df

code                                          name
0     8                             Human development
1     1                           Economic management
2     6         Social protection and risk management
3     5                         Trade and integration
4     2                      Public sector governance
5    11  Environment and natural resources management
6     7                   Social dev/gender/inclusion
7     4      Financial and private sector development
8    10                             Rural development
9     9                             Urban development
10    3                                   Rule of law

We can see that there are 11 major project themes and corresponding codes. We can use this to create a dataframe with no missing names by merging the two dataframes.

In [22]:
# merge dataframes
merged_df=pd.merge(mjtheme_df, unique_mjtheme_df, on='code',suffixes=('_x','')).drop('name_x', axis=1)
merged_df.head()

code               name
0    8  Human development
1    8  Human development
2    8  Human development
3    8  Human development
4    8  Human development

In [34]:
# checking the dataframe doesn't have anymore missing values
merged_df[merged_df.name==''].count()

code    0
name    0
dtype: int64

2. Next method is to sort values by code and names, fill blanks with NaN and use backward fill to fill names with appropriate names

In [26]:
# sort the rows by code and missing values


sort_theme= mjtheme_df.sort_values(['code', 'name'], ascending=True)
sort_theme.head(10)

code                 name
212     1                     
363     1                     
1024    1                     
1114    1                     
1437    1                     
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management

In [35]:
#fill missing values with NaNs 
sort_theme[sort_theme.name=='']= np.nan
sort_theme.head(10)

code                 name
212   NaN                  NaN
363   NaN                  NaN
1024  NaN                  NaN
1114  NaN                  NaN
1437  NaN                  NaN
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management

In [30]:
filled_df= sort_theme.fillna(method='bfill')
filled_df

code                 name
212     1  Economic management
363     1  Economic management
1024    1  Economic management
1114    1  Economic management
1437    1  Economic management
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management
220     1  Economic management
222     1  Economic management
223     1  Economic management
249     1  Economic management
357     1  Economic management
453     1  Economic management
454     1  Economic management
458     1  Economic management
497     1  Economic management
647     1  Economic management
648     1  Economic management
784     1  Economic management
803     1  Economic management
841     1  Economic management
900     1  Economic management
1010    1  Economic management
1045    1  Economic management
1056    1  Economic management
1057    1  Economic management
1078    1  Economic management
...   ...                  ...
597     9    Urban development
598     9    Urban development
650     9    Urban development
669     9    Urban development
732     9    Urban development
737     9    Urban development
738     9    Urban development
739     9    Urban development
742     9    Urban development
822     9    Urban development
917     9    Urban development
930     9    Urban development
940     9    Urban development
962     9    Urban development
1095    9    Urban development
1102    9    Urban development
1103    9    Urban development
1104    9    Urban development
1166    9    Urban development
1290    9    Urban development
1303    9    Urban development
1318    9    Urban development
1342    9    Urban development
1347    9    Urban development
1425    9    Urban development
1426    9    Urban development
1428    9    Urban development
1470    9    Urban development
1473    9    Urban development
1495    9    Urban development

[1499 rows x 2 columns]

In [31]:
filled_df.reset_index().drop('index', axis=1)

code                 name
0       1  Economic management
1       1  Economic management
2       1  Economic management
3       1  Economic management
4       1  Economic management
5       1  Economic management
6       1  Economic management
7       1  Economic management
8       1  Economic management
9       1  Economic management
10      1  Economic management
11      1  Economic management
12      1  Economic management
13      1  Economic management
14      1  Economic management
15      1  Economic management
16      1  Economic management
17      1  Economic management
18      1  Economic management
19      1  Economic management
20      1  Economic management
21      1  Economic management
22      1  Economic management
23      1  Economic management
24      1  Economic management
25      1  Economic management
26      1  Economic management
27      1  Economic management
28      1  Economic management
29      1  Economic management
...   ...                  ...
1469    9    Urban development
1470    9    Urban development
1471    9    Urban development
1472    9    Urban development
1473    9    Urban development
1474    9    Urban development
1475    9    Urban development
1476    9    Urban development
1477    9    Urban development
1478    9    Urban development
1479    9    Urban development
1480    9    Urban development
1481    9    Urban development
1482    9    Urban development
1483    9    Urban development
1484    9    Urban development
1485    9    Urban development
1486    9    Urban development
1487    9    Urban development
1488    9    Urban development
1489    9    Urban development
1490    9    Urban development
1491    9    Urban development
1492    9    Urban development
1493    9    Urban development
1494    9    Urban development
1495    9    Urban development
1496    9    Urban development
1497    9    Urban development
1498    9    Urban development

[1499 rows x 2 columns]

In [32]:
# checking the dataframe doesn't have anymore missing values
filled_df[filled_df.name==''].count()

code    0
name    0
dtype: int64

#### 2. Top 10 major project themes

In [33]:
filled_df['name'].value_counts()

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Rule of law                                      15
Name: name, dtype: int64